core/os/device/ggp/device.go

```diff
@@ -0,0 +1,285 @@
+			continue
+		}
+		if sshInitErrBuf.Len() != 0 {
+			log.W(ctx, "'ggp ssh init --gamelet %v -s' finished with error: %v", inf[1], sshInitErrBuf.String())
+			continue
+		}
+		envs := []string{
+			"YETI_DISABLE_GUEST_ORC=1",
+			"YETI_DISABLE_STREAMER=1",
+			"YETI_INTERNAL_DISABLE_USER_SAVE_DRIVER=1",
+		}
+		cfg := remotessh.Configuration{
+			Name: inf[0],
+			Env:  envs,
+		}
+		if err := json.Unmarshal(sshInitOutBuf.Bytes(), &cfg); err != nil {
+			log.W(ctx, "Failed at unmarshaling 'ggp ssh init --gamelet %v -s' output, fallback to use default config, err: %v", inf[1], err)
+		}
+		configs = append(configs, GGPConfiguration{cfg, inf[1]})
+	}
+	return configs, nil
+}
+
+// Monitor updates the registry with devices that are added and removed at the
+// specified interval. Monitor returns once the context is cancelled.
+func Monitor(ctx context.Context, r *bind.Registry, interval time.Duration) error {
+	unlisten := registry.Listen(bind.NewDeviceListener(r.AddDevice, r.RemoveDevice))
+	defer unlisten()
+
+	for _, d := range registry.Devices() {
+		r.AddDevice(ctx, d)
+	}
+
+	var lastErrorPrinted time.Time
+
+	for {
+		if err := func() error {
+			configs, err := getConfigs(ctx)
+			if err != nil {
+				return err
+			}
+			if err := scanDevices(ctx, configs); err != nil {
+				if time.Since(lastErrorPrinted).Seconds() > printScanErrorsEveryNSeconds {
+					log.E(ctx, "Couldn't scan devices: %v", err)
+					lastErrorPrinted = time.Now()
+				}
+			} else {
+				lastErrorPrinted = time.Time{}
+			}
+			return nil
+		}(); err != nil {
+			return err
+		}
+
+		select {
+		case <-task.ShouldStop(ctx):
+			return nil
+		case <-time.After(interval):
+		}
+	}
+}
+
+// Devices returns the list of attached GGP devices.
+func Devices(ctx context.Context) ([]bind.Device, error) {
+	configs, err := getConfigs(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := scanDevices(ctx, configs); err != nil {
+		return nil, err
+	}
+	devs := registry.Devices()
+	out := make([]bind.Device, len(devs))
+	for i, d := range devs {
+		out[i] = d
+	}
+	return out, nil
+}
+
+func deviceStillConnected(ctx context.Context, d *Binding) bool {
+	return d.Status(ctx) == bind.Status_Online
+}
+
+func scanDevices(ctx context.Context, configurations []GGPConfiguration) error {
+	cacheMutex.Lock()
+	defer cacheMutex.Unlock()
+	allConfigs := make(map[string]bool)
+
+	for _, cfg := range configurations {
+		allConfigs[cfg.Name] = true
+
+		// If this device already exists, see if we
+		// can/have to remove it
+		if cached, ok := cache[cfg.Name]; ok {
+			if !deviceStillConnected(ctx, cached) {
+				delete(cache, cfg.Name)
+				registry.RemoveDevice(ctx, cached)
+			}
+		} else {
+			if device, err := remotessh.GetConnectedDevice(ctx, cfg.Configuration); err == nil {
+				dev := Binding{
+					Device:  device,
+					Gamelet: cfg.Gamelet,
+				}
+				device.Instance().Configuration.OS.Kind = bd.Stadia
```

Comment:
This is a bit hacky. Is there any way for the device itself to know it is Stadia? Is there any chance the device ID will be broken? (it used to be a hash of the fields of the device, not sure if that's still the case)

Reply:
So how we handle Stadia, we may want to trace locally, at which point it is treated the same as a linux machine. (I.e. push gapis/gapit to the device).